uboot-mediatek: add support for Zyxel EX5601-T0 with custom partitions

ftp21 opened the pull request:
The flash procedure is similar to the Xiaomi AX6000 router.

Load openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb from original Zyxel U-Boot:
```
tftpboot openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb
bootm 0x46000000
```

Load mtd-rw
`insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1`

Format ubi and create ubootenv partitions
```
ubidetach -p /dev/mtd5; ubiformat /dev/mtd5 -y; ubiattach -p /dev/mtd5
ubimkvol /dev/ubi0 -n 0 -N ubootenv -s 128KiB
ubimkvol /dev/ubi0 -n 1 -N ubootenv2 -s 128KiB
```

Copy openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb to /tmp and create recovery partition.
If your recovery image is larger than 10MiB, size the recovery partition accordingly to make it fit.
```
ubimkvol /dev/ubi0 -n 2 -N recovery -s 10MiB
ubiupdatevol /dev/ubi0_2  openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb 
```


Copy preloader and uboot to /tmp and write them in the mtd
```
mtd write /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-preloader.bin bl2
mtd write /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-bl31-uboot.fip fip
```

Now write the firmware:
`sysupgrade -n  /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-squashfs-sysupgrade.itb`

To create a correct BL2, I had to add a profile for 'spim:4k+256' as I could not find a way to value the variable 'NAND_TYPE'.

hauke commented:
Please restructure the commits to make review easier.
1. Create common dtsi file
2. Add the new version of the board

Please add the Signed-off-by:  line at the end of the commit description.

dangowrt commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0-ubootmod.dts:
```diff
@@ -0,0 +1,97 @@
+/ {
+	model = "Zyxel EX5601-T0 ubootmod";
+	compatible = "zyxel,ex5601-t0-ubootmod", "mediatek,mt7986a";
+};
```
Since https://github.com/openwrt/openwrt/commit/572ea680705311a5ba08cb0f5da4c7dc2b792fd3#diff-0c89bd92e2a03d26907d885e84652e23bed36ef4ffcdbb1363d696c3e85b33b6 it is mandatory to include a `memory` node in device tree.
```suggestion
	memory@40000000 {
		device_type = "memory";
		reg = <0x40000000 0x20000000>;
	};
};
```

dangowrt commented on package/boot/arm-trusted-firmware-mediatek/Makefile:
```diff
@@ -174,6 +174,15 @@ define Trusted-Firmware-A/mt7986-spim-nand-ddr4
   NAND_TYPE:=spim:2k+64
 endef
 
+define Trusted-Firmware-A/mt7986-spim-nand-ddr4-4k
```
`mt7986-spim-nand-4k-ddr4` as a name would make more sense to me, because the `4k` apply to `spim-nand` and not to `ddr4`

dangowrt reviewed (changes requested):
Please squash everything two commits:

First commit adding the build for this device to the `uboot-mediatek` package

Second commit adding the `ubootmod` variant in `target/linux/mediatek/image/filogic.mk`

The changes themselves look good now.

dangowrt commented on package/boot/uboot-mediatek/Makefile:
```diff
@@ -309,6 +309,18 @@ define U-Boot/mt7986_xiaomi_redmi-router-ax6000
+  BL2_BOOTDEV:=spim-nand
+  BL2_SOC:=mt7986
+  BL2_DDRTYPE:=ddr4
+  DEPENDS:=+trusted-firmware-a-mt7986-spim-nand-ddr4-4k
```
You will also have to change the name here

dangowrt commented on package/boot/arm-trusted-firmware-mediatek/Makefile:
```diff
@@ -237,7 +246,8 @@ TFA_TARGETS:= \
 	mt7986-snand-ddr4 \
-	mt7986-spim-nand-ddr4
+	mt7986-spim-nand-ddr4 \
+	mt7986-spim-nand-ddr4-4k
```
Also change the name here.

dangowrt commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -421,5 +421,39 @@ define Device/zyxel_ex5601-t0-stock
+  IMAGE/sysupgrade.itb := append-kernel | \
+        fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb external-static-with-rootfs | append-metadata
+  ARTIFACTS := preloader.bin bl31-uboot.fip
+  ARTIFACT/preloader.bin := mt7986-bl2 spim-nand-ddr4-4k
```
Also here still the old filename

ftp21 commented:
Add mandatory memory section also on stock dts.
Fix "01_leds" also for stock.

dangowrt commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -421,5 +421,39 @@ define Device/zyxel_ex5601-t0-stock
+
+define Device/zyxel_ex5601-t0-ubootmod
+  DEVICE_VENDOR := Zyxel
+  DEVICE_MODEL := EX5601-T0  (ubootmod)
```
use `DEVICE_VARIANT` maybe?

ftp21 replied:
DEVICE_VARIANT used on ubootmod and stock layout

ftp21 commented:
Are there any other changes to be made to commits?

hauke commented:
Please rebase this pull request, if has conflicts with master.

ftp21 committed:
uboot-mediatek: add initial Zyxel EX5601-T0 support

Flash procedure is described in next commit.

TLDR:
Copy preloader and uboot to /tmp and write them in the mtd.
This will also require new UBI partition and
 volumes to boot openwrt.

mtd write /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-preloader.bin bl2
mtd write /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-bl31-uboot.fip fip

Changelist:
 - Added profile for 4k+256 SPI NAND_TYPE
 - Added basic Zyxel EX5601-T0 uboot profile

Backported from hitech95 branch:
 - Button RESET pin fix
 - Button WPS pin fix

Signed-off-by: Valerio 'ftp21' Mancini <ftp21@ftp21.eu>
Signed-off-by: Nicolò Veronese <nicveronese@gmail.com>

ftp21 commented:
> Please rebase this pull request, if has conflicts with master.

Rebased to resolve conflicts

hitech95 commented:
Aren't aliases for the leds (boot) and gmac missing?

ftp21 committed:
mediatek: add Zyxel EX5601-T0 with uboot custom partition

The flash procedure is similar to the Xiaomi AX6000 router.

Load openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb from original Zyxel U-Boot:

tftpboot openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb
bootm 0x46000000
Load mtd-rw
insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1

Format ubi and create ubootenv partitions

ubidetach -p /dev/mtd5; ubiformat /dev/mtd5 -y; ubiattach -p /dev/mtd5
ubimkvol /dev/ubi0 -n 0 -N ubootenv -s 128KiB
ubimkvol /dev/ubi0 -n 1 -N ubootenv2 -s 128KiB
Copy openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb to /tmp and create recovery partition.
If your recovery image is larger than 10MiB, size the recovery partition accordingly to make it fit.

ubimkvol /dev/ubi0 -n 2 -N recovery -s 10MiB
ubiupdatevol /dev/ubi0_2  openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-initramfs-recovery.itb
Copy preloader and uboot to /tmp and write them in the mtd

mtd write /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-preloader.bin bl2
mtd write /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-bl31-uboot.fip fip
Now write the firmware:
sysupgrade -n  /tmp/openwrt-mediatek-filogic-zyxel_ex5601-t0-ubootmod-squashfs-sysupgrade.itb

To create a correct BL2, I had to add a profile for 'spim:4k+256' as I could not find a way to value the variable 'NAND_TYPE'.

Features and fixes from hitech95 tree has been squashed, I'm attaching his commit message:

The Power LED was not working correctly and not reacting
 to the boot process and statuses.

The board has space (footprint) for an unpopulated Zigbee chip,
 while we dont know the device model having this chip populated
 we have to assure that the common dts doesnt enable
 interfaces that share pins with such device.

In this instance the PCIe and the uart1 and uart2 are disabled.
Some of the control PCIE pins seems to be used for the Zigbee chip,
 UART1 seems to be used as a flash port while UART2 should be the
 main comunication interface of Zigbee chip.

The Zigbee chip should be a EFR32MG21. But the pins used for UART
 seems to be not on standard PINS used by other adapters.
 So it cannot run firmwares shared on the web.
 But it should be possible to build a custom firmware with
 the corrtect pinmux.

This commit also contains the following squashed commit from hitech95
 - mediatek: fix sysupgrade for Zyxel EX7601-T0 ubootmod

Changes and fixes added in common board:
 - added aliases for boot status leds.
 - added aliases for the mac-label-device.
 - added pin claims for core features (MDIO and UART 0)
 - added default LEDs configuration (01_leds)
 - added default network configuration (02_network)
 - added missing kmod-usb3 module for USB3
 - fixed LED names
 - fixed reset pin for SLIC chip
 - removed unused pinmux configurations and devices
 - fix LAN (switch) port numbering
 - using nvmem cells for wifi eeprom, dropping deprecated "mediatek,mtd-eeprom"
 - proper factory partition and mac address handling
 - cleaned up spi_nand sections and partition

Changes and fixxes added in stock layout:
 - added NMBM, if u-boot has it, the kernel must be informed.

Co-authored-by: Nicolò Veronese <nicveronese@gmail.com>
Co-developed-by: Nicolò Veronese <nicveronese@gmail.com>
Signed-off-by: Nicolò Veronese <nicveronese@gmail.com>
Signed-off-by: Valerio 'ftp21' Mancini <ftp21@ftp21.eu>

hitech95 commented:
Rebse on master, this depends on #13599 as we have found a issue with the SPIM controller.
This is the base support for the device another PR with more functionality will be open once completed and tested.

pameruoso reviewed (approved):
Everything looks good to me.

hitech95 commented:
#13599 has been merged so rebase on top of master

ftp21 commented:
Hello, 
News about review?

hitech95 committed:
uboot-envtools: add support for Zyxel EX5601-T0 ubootmod

The ubootmod bootlaoder for EX5601-T0 uses two partitions
 in ubi to store enviroment variables. so proper config
 is needed.

Signed-off-by: Nicolò Veronese <nicveronese@gmail.com>

hitech95 commented:
@hauke any changes left changes that are required?

Crav88 commented:
Hello,
News about review?

Crav88 commented:
mmm will be nice have a list of conflict that must be resolved :D @dangowrt

mo3rfan commented on package/boot/uboot-mediatek/patches/439-add-zyxel_ex5601-t0.patch:
```diff
@@ -0,0 +1,431 @@
++
++			partition@380000 {
++				label = "fip";
++				reg = <0x380000 0x0200000>;
```
@ftp21 Why is `zloader` starting at `0x540000` if FIP starts at`0x380000` and size of FIP is `0x0200000`?

Is the size of FIP a typo? In the stock firmware, the FIP size is `0x01C0000`. 

Boot log from stock firmware:

```
[    1.098413] Creating 8 MTD partitions on "nmbm_spim_nand":
[    1.103885] 0x000000000000-0x000000100000 : "BL2"
[    1.109000] 0x000000100000-0x000000180000 : "u-boot-env"
[    1.114611] 0x000000180000-0x000000380000 : "Factory"
[    1.119938] 0x000000380000-0x000000540000 : "FIP"
[    1.124915] 0x000000540000-0x000000580000 : "zloader"
[    1.130233] 0x000000580000-0x000004580000 : "ubi"
[    1.135259] 0x000004580000-0x000008580000 : "ubi2"
[    1.140358] 0x000008580000-0x00001e000000 : "zyubi"
```

ftp21 replied:
This should be a typo. 
I would like to work on the DTS refactor for Zyxel EX5600/T56,WX5600,EX5512,WX5512 like original sources and i need help :)
